posthog/api/feature_flag.py

```diff
@@ -113,8 +113,135 @@ def allow_request(self, request, view):
         return super().allow_request(request, view)
 
 
+class EvaluationTagSerializerMixin(serializers.Serializer):
+    """
+    Serializer mixin that handles evaluation tags for feature flags.
+    Evaluation tags mark which organizational tags also serve as runtime evaluation constraints.
+
+    Note: SDK clients must send 'evaluation_environments' in their flag evaluation requests
+    for these constraints to take effect. Without this parameter, all flags are evaluated
+    regardless of their evaluation tags.
+    """
+
+    evaluation_tags = serializers.ListField(required=False, write_only=True)
+
+    def validate(self, attrs):
+        """Validate that evaluation_tags are a subset of tags.
+
+        This ensures that evaluation tags (which control runtime evaluation)
+        are always a subset of organizational tags. This maintains the conceptual
+        model where evaluation tags are tags that ALSO serve as constraints.
+        """
+        attrs = super().validate(attrs)
+
+        # Only validate if we have initial_data (not during partial updates without these fields)
+        if not hasattr(self, "initial_data"):
+            return attrs
+
+        # Get evaluation_tags from the request
+        evaluation_tags = self.initial_data.get("evaluation_tags")
+
+        # Only validate if evaluation_tags are provided and non-empty
+        # Note: evaluation_tags=[] is valid (clears all evaluation tags)
+        if evaluation_tags is not None and evaluation_tags:
+            from posthog.api.tagged_item import tagify
+
+            # Get tags from initial_data, defaulting to empty list if not provided
+            # Important: We validate against the raw request data, not processed attrs,
+            # because TaggedItemSerializerMixin handles tags separately
+            tags = self.initial_data.get("tags", [])
+
+            # Normalize both lists using tagify for consistent comparison
+            # tagify handles case normalization and special characters
+            normalized_tags = {tagify(t) for t in tags} if tags else set()
+            normalized_eval_tags = {tagify(t) for t in evaluation_tags}
```

Comment:
this does make flag updates more expensive whenever we update flags with tags. It's a small use case, but wanted to call it out as a potential (but unlikely bottleneck)